### ISSUE TITLE
RandomRain never paints the last row or column of the image (start coordinate scaled by H - h - 1)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4604]
## Summary

`RandomRain` can never paint the last row or the last column of the image, unless the drop size is exactly one short of the image on that axis. The start coordinate of each drop is scaled by `H - h - 1` (and `W - w - 1`), one less than the number of admissible start positions, so after the `long` truncation the last start row is `H - h - 2` and the drop's last painted row is at most `H - 2`.

## Reproduction

```python
import torch
import kornia.augmentation as K

rows, cols = set(), set()
for seed in range(300):
    torch.manual_seed(seed)
    out = K.RandomRain(number_of_drops=(20, 20), drop_height=(1, 1), drop_width=(0, 0), p=1.0)(torch.zeros(1, 3, 6, 10))
    nz = (out[0, 0] != 0).nonzero()
    rows |= set(nz[:, 0].tolist()); cols |= set(nz[:, 1].tolist())
print(sorted(rows), sorted(cols))
```

Output on torch 2.14.0, CPU (main at `90650596d`):

```
[0, 1, 2, 3] [0, 1, 2, 3, 4, 5, 6, 7, 8]
```

Six thousand single-pixel drops on a `6 x 10` image never land on rows 4 or 5 or on column 9. With `drop_height=(2, 2)`, `drop_width=(-2, -2)` the painted rows are `0..4` and the columns `0..8`: row 5 and column 9 are still never touched.

## Cause

`kornia/augmentation/_2d/intensity/random_rain.py`, `apply_transform`:

```python
random_y_coords = coordinates_of_drops[:, 0] * (image.shape[2] - height_of_drop - 1)
if width_of_drop > 0:
    random_x_coords = coordinates_of_drops[:, 1] * (image.shape[3] - width_of_drop - 1)
else:
    random_x_coords = coordinates_of_drops[:, 1] * (image.shape[3] + width_of_drop - 1) - width_of_drop
coords = torch.cat([...]).to(image.device, dtype=torch.long)
```

`coordinates_of_drops` is uniform on `[0, 1)`, so `u * (H - h - 1)` truncated to `long` is at most `H - h - 2`. A drop of height `h >= 2` spans `h + 1` rows (documented in the class), so its last row is at most `H - 2`; a single-pixel drop (`h = 1`) starts at most at `H - 3`. The only case that reaches the far edge is `h = H - 1`, where the multiplier is `0` and every drop starts at row `0`.

The expected behaviour is that a drop can start anywhere in `[0, H - 1 - h]`, i.e. the multiplier should be `H - h` (the count of admissible start positions), with the same change on the width axis in both branches.

## Scope

This is the *placement* of the drops. #4567 is a different defect: the truncation of the *size* draws, which makes the upper bound of `drop_height` / `drop_width` / `number_of_drops` practically never drawn. Both truncations are in the same class but in different tensors.

Found while reviewing #4566, which documents the current placement behaviour with a pin that names this issue.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
